pkg/codegen/go/gen_program.go

```diff
@@ -1546,3 +1556,162 @@ func programPackageDefs(program *pcl.Program) ([]*schema.Package, error) {
 	}
 	return defs, nil
 }
+
+// fileImporter tracks imports in a single generated file.
+// It ensures that there are no conflicts between imports
+// with the same package name.
+type fileImporter struct {
+	// used holds all identifier names that have been used
+	// for imports in this file, and the imports that they refer to.
+	used map[string]string // identifier name -> import path
+
+	// For import paths where the package name is not unique,
+	// this map holds the name that was used for the import.
+	aliases map[string]string // import path -> alias
+}
+
+func newFileImporter() *fileImporter {
+	return &fileImporter{
+		used:    make(map[string]string),
+		aliases: make(map[string]string),
+	}
+}
+
+// Import imports a package with the given import path
+// and returns the name that should be used to refer to it.
+//
+// name is the name of the package at the import path.
+// This must always match the 'package' statement in the imported package.
+//
+// Note that returned name may be different from the name argument.
+// This happens when the name is already used for another import,
+// and requires an alias to avoid a conflict.
+//
+//	foo := i.Import("example.com/foo", "foo")
+//	if foo == "foo" {
+//	    fmt.Printf(`import "example.com/foo"`)
+//	} else {
+//	    fmt.Printf(`import %s "example.com/foo"`, foo)
+//	}
+//
+// Import will never return the same name for two different import paths
+// in the same file.
+//
+// For example:
+//
+//	aws1 := i.Import("example.com/foo/aws", "aws")
+//	aws2 := i.Import("example.com/bar/aws", "aws")
+//	fmt.Println(aws1 == aws2) // false
+func (fi *fileImporter) Import(importPath string, name string) (actualName string) {
+	contract.Requiref(importPath != "", "importPath", "must not be empty")
+	contract.Requiref(name != "", "name", "must not be empty (importPath: %q)", importPath)
+
+	// For readability, always add an alias if the package name
+	// does not match the base name of the import path.
+	// For example, "example.com/foo-go" with package "foo"
+	// should get:
+	//
+	//	import foo "example.com/foo-go"
+	if filepath.Base(importPath) != name {
+		defer func() { fi.aliases[importPath] = actualName }()
+	}
+
+	// Already imported using the same name.
+	if imported, ok := fi.used[name]; ok && imported == importPath {
+		return name // no alias
+	}
+
+	// Preferred name has not yet been used.
+	if _, ok := fi.used[name]; !ok {
+		fi.used[name] = importPath
+		return name
+	}
+
+	// Already imported with an alias. Use that alias.
+	if other, ok := fi.aliases[importPath]; ok {
+		return other
+	}
+
+	// Find a unique name to use as an alias.
```

Comment:
Can we try using some of the importPath for this first? So like in the regress example added in this PR we'd get "awsxEcs" rather than "ecs2"? e.g. ("example.com/bar/aws", "aws") -> "barAws".

Reply:
Alright, maybe a "join last two components, and if that's unique, use that, otherwise numbered"? Yeah, I can see that being more readable, though it would be "awsx_ecs" or "awsxecs"; package names aren't camelCase (and even `_` is suspect, but I could bend on that for this).

Reply:
Done! We now try to build a unique `foo_bar` from the last two components of the import path before falling back to the numerical suffix.